databallpy/features/covered_distance.py

```diff
@@ -0,0 +1,473 @@
+
+            # smoothing the signal
+            if filter_type is not None:
+                gradient_x = _filter_data(
+                    gradient_x,
+                    filter_type=filter_type,
+                    window_length=window,
+                    polyorder=poly_order,
+                )
+                gradient_y = _filter_data(
+                    gradient_y,
+                    filter_type=filter_type,
+                    window_length=window,
+                    polyorder=poly_order,
+                )
+
+            res_dict[column_id + f"_{new_name[0]}x"] = np.array(gradient_x)
+            res_dict[column_id + f"_{new_name[0]}y"] = np.array(gradient_y)
+            res_dict[column_id + f"_{new_name}"] = np.linalg.norm(
+                [gradient_x, gradient_y], axis=0
+            )
+
+        new_columns_df = pd.DataFrame(res_dict)
+
+        if inplace:
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", category=pd.errors.PerformanceWarning)
+                df[new_columns_df.columns] = new_columns_df
+            return None
+        else:
+            return pd.concat([df, new_columns_df], axis=1)
+
+    except Exception as e:
+        LOGGER.exception(f"Found unexpected exception in _differentiate: \n{e}")
+        raise e
+
+
+#%% 3-7-2024
+def _parse_intervals(intervals):
+    if not all(isinstance(element, (int, float)) for element in intervals):
+        raise TypeError("All elements in the tuple must be integers or floats")
+    
+    if len(intervals) % 2 != 0:
+        raise ValueError("Intervals must contain an even number of elements.")
+    
+    return [(min(intervals[i], intervals[i+1]), max(intervals[i], intervals[i+1])) for i in range(0, len(intervals), 2)]
+
+def get_covered_distance(
+    tracking_data: pd.DataFrame,
+    player_ids: list[str],
+    framerate: int,
+    velocity_intervals: tuple[float, ...] | None = None,
+    acceleration_intervals: tuple[float, ...] | None = None
+    ) -> dict:
+    """Function that calculates the distance covered in specified velocity and acceleration intervals
+
+    Args:
+        tracking_data:      tracking data
+        player_ids:         list containing players. Example = ['home_8', 'away_3', 'home_16']
+        framerate:          sample frequency in Hertz
+        velocity_intervals: tuple that contains the velocity interval(s). For example [0 3] & [4 8]: (0,3,8,4)
+        acc_intervals:      tuple that contains the acceleration interval(s). For example [-3 0] & [3 5]: (-3,0,3,5)      
+
+    Returns:
+        dict: a dictionary with for every player the total distance covered and optionally the 
+        distance covered within the given velocity and/or acceleration threshold(s)
+    
+    Notes: 
+        The function requires the velocity for every player calculated with the add_velocity function. 
+        The acceleration for every player depends on the presence of acceleration intervals in the input
```

Comment:
Not true, if they only want covered distance velocities are not necessary. Maybe mention that `add velocity` and `add_acceleration` are required if velocity interval / acceleration interval is not None. 

If velocity is not in the tracking data and they only want total covered distance, then calculate the raw distance covered per frame and use that for total distance calculations. From a users perspective, it makes no sense to have velocity in the data while you only want covered distance.

Reply:
I use the velocity calculated with get_velocity function to get the covered distance (s = v * t). So in this function I also need velocity for the total distance. Do you want me to change this to an Euclidean distance?

Reply:
I think the get_velocity function is always required, because I use the velocity to calculate the covered distance (s = v * t). Or do you want me to change this to Euclidean distance?

Reply:
Okay I have given it some thought and I think you are right, we need to be very clear that we calculate the covered distance by the velocity magnitude of each frame and let the user call the `add_velocity` function before being able to call this function.

Just to explain, I thought it would make no sense, from a users perspective, to make it mandatory to have velocity magnitudes in the tracking data if they only wish to get the covered distance. But making the calculations in the function itself (if velocity is not in the given tracking data), would probably make things less logical. For instance, if a user calculates the covered distance (without velocity in the tracking data), than computes the velocity, and then again calculates the covered distance (with the velocity), the outcome of the function would be different. This would be very illogical for a user. So therefore, it is a good idea to be very clear that we calculate covered distance based on the velocity magnitude, and that it is required to call this functions. Thanks for pointing that out.